[DOC-2269] Document env vars and init scripts for Docker single-node clusters (#15099)
cockroachdb/docs: _includes/v22.2/start-in-docker/mac-linux-steps-single-node.md

When you use the `cockroach start-single-node` command to start a single-node cluster with Docker, additional features are available to help with testing and development.

{{site.data.alerts.callout_danger}}
Single-node clusters are not highly-available or fault-tolerant. They are not appropriate for production use.
{{site.data.alerts.end}}

- You can optionally set the following [Docker environment variables](https://docs.docker.com/engine/reference/commandline/run/#set-environment-variables--e---env---env-file) to create a database and user automatically and to set a password for the user.
  - `COCKROACH_DATABASE`
  - `COCKROACH_USER`
  - `COCKROACH_PASSWORD`

>   To prevent data loss, the environment variables are ignored if the `/cockroach/cockroach-data` directory within the container is not empty. To re-initialize the default database, user, and password from the environment variable values, delete the contents of `/cockroach/cockroach-data` within the running container.

Reviewer: nit: the environment variables and initialization scripts in `docker-entrypoint-initdb.d` will be ignored

> - You can optionally mount a directory of initialization scripts into the `docker-entrypoint-initdb.d` directory within the container. These scripts are run after CockroachDB starts and after the database and user (if specified) have been created. The scripts run in the alphanumeric sort order imposed by your locale.

Reviewer: nit: and after the database and user (if specified via the environment variables)

1. Start the cluster node.

    The following command starts a single-node cluster that:
 
    - Stores its data in the `roach-single` volume on the Docker host, which is mounted on the `/cockroach/cockroach-data` directory within the container.
    - If the `/cockroach/cockroach-data` within the container is empty, the specified database, user, and password are created automatically. Instead of specifying each value directly by using the `-e` or `--env` flag, you can store them in a file on the Docker host. Use one key-value pair per line and set the `--env-file` flag to the file's path.
    - Bind-mounts the `~/init-scripts` directory on the Docker host onto the `/docker-entrypoint-initdb.d` directory within the container. Initialization scripts stored in this directory are run after CockroachDB starts and the default database, user, and password are initialized.
    - Accepts database client connections on IP address 172.18.0.3 (which resolves to the hostname `roach-single` within the Docker container) on port 26257.
    - Accepts connections to the DB Console on IP address 172.18.0.3 (which resolves to the hostname `roach-single` within the Docker container) on port 8080.

> 

Reviewer: As illustrated in [the existing docker page](https://www.cockroachlabs.com/docs/stable/start-a-local-cluster-in-docker-mac.html#step-1-create-a-bridge-network), we need to create a docker network first.
In the following example, we have the IP and advertise address `172.18.0.3`, so we need to do
```
docker network create -d bridge --subnet=172.18.10.3/16 roachnet
```
Note that the subnet address can be customized.

Author: Since the existing doc says "Since you'll be running multiple Docker containers on a single host, with one CockroachDB node per container, you need to create what Docker refers to as a [bridge network](https://docs.docker.com/engine/userguide/networking/#/a-bridge-network)", I made the assumption that you don't need a Docker network, since you aren't running multiple Docker containers on a single host. Do I need to update the non-single-user doc to reflect a different reason that the Docker network is needed?

Reviewer: You're right, we don't need to create a network in this case 👍

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    docker run -d \
              --env COCKROACH_DATABASE={DATABASE_NAME} \
              --env COCKROACH_USER={USER_NAME} \
              --env COCKROACH_PASSWORD={PASSWORD} \
              --name=roach-single \
              --hostname=roach-single \
              --ip=172.18.0.3 \

>               --advertise-addr=172.18.0.3 \

Reviewer: There's no `--advertise-addr` flag for docker run, so I think we should remove it.

Author: Done.

              -p 26257:26257 -p 8080:8080 \
              -v "roach-single:/cockroach/cockroach-data" \
              -v "~/init-scripts:/docker-entrypoint-initdb.d" \
              cockroachdb/cockroach:latest start-single-node
    ~~~

    By default, a `certs` directory is created and CockroachDB starts in secure mode. To prevent this, add the `--insecure` flag after the `start-single-node` sub-command.

    {{site.data.alerts.callout_info}}
    The `COCKROACH_DATABASE`, `COCKROACH_USER`, and `COCKROACH_PASSWORD` environment variables and the contents of the `/docker-entrypoint-initdb.d` directory are ignored if you use `cockroach start` rather than `cockroach start-single-node`. They are also ignored if the `/cockroach/cockroach-data` directory within the container is not empty.
    {{site.data.alerts.end}}

1. Perform a one-time initialization of the cluster:

    {% include_cached copy-clipboard.html %}
    ~~~ shell

>     docker exec -it roach-single ./cockroach init --insecure

Reviewer: I don't think with `start-single-node`, we will need this step. `cockroach init` is usually used for multi-node cluster.

    ~~~

    You'll see the following message:

    ~~~
    Cluster successfully initialized
    ~~~

    At this point, the cluster node also prints helpful [startup details](cockroach-start.html#standard-output) to its log. For example, the following command retrieves node 1's startup details:

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    docker exec -it roach-single grep 'node starting' cockroach-data/logs/cockroach.log -A 11
    ~~~

    The output will look something like this:

>     ~~~

Reviewer: I think this output is for an insecure node, but not for a secure node as in the `Start the cluster node` example. 

The output for a secure node (for v22.2.0-beta.1) seems to be 
```
CockroachDB node starting at 2022-09-28 14:35:27.495508396 +0000 UTC m=+1.370085757 (took 0.5s)
build:               CCL v22.2.0-beta.1 @ 2022/09/26 18:49:07 (go1.19.1)
webui:               https://127.0.0.1:8080
sql:                 postgresql://root@127.0.0.1:26257/defaultdb?sslcert=certs%2Fclient.root.crt&sslkey=certs%2Fclient.root.key&sslmode=verify-full&sslrootcert=certs%2Fca.crt
sql (JDBC):          jdbc:postgresql://127.0.0.1:26257/defaultdb?sslcert=certs%2Fclient.root.crt&sslkey=certs%2Fclient.root.key&sslmode=verify-full&sslrootcert=certs%2Fca.crt&user=root
RPC client flags:    /cockroach/cockroach <client cmd> --host=127.0.0.1:26257 --certs-dir=certs
logs:                /cockroach/cockroach-data/logs
temp dir:            /cockroach/cockroach-data/cockroach-temp2611102055
external I/O path:   /cockroach/cockroach-data/extern
store[0]:            path=/cockroach/cockroach-data
storage engine:      pebble
clusterID:           60f29a4e-1c87-4b0c-805d-eb73460766b1
status:              initialized new cluster
nodeID:              1
```

Author: This output also makes me question the need to create a Docker network, since you are reaching the cluster at `127.0.0.1:26257`. Are these in conflict?

Author: I updated the command like this and I added a bit of text explaining that Docker adds a DNS rule (I think the DNS rule is based on the `--name` flag actually, and that the `--hostname` flag should be used to change the hostname within the container if it should be different from `--name`. Does CRDB require setting both?)

    CockroachDB node starting at {{ now | date: "%Y-%m-%d %H:%M:%S.%6 +0000 UTC" }}
    build:               CCL {{page.release_info.version}} @ {{page.release_info.build_time}} (go1.12.6)
    webui:               http://roach-single:8080
    sql:                 postgresql://root@roach-single:26257?sslmode=disable
    client flags:        /cockroach/cockroach <client cmd> --host=roach-single:26257
    logs:                /cockroach/cockroach-data/logs
    temp dir:            /cockroach/cockroach-data/cockroach-temp273641911
    external I/O path:   /cockroach/cockroach-data/extern
    store[0]:            path=/cockroach/cockroach-data
    status:              initialized new cluster
    clusterID:           1a705c26-e337-4b09-95a6-6e5a819f9eec
    nodeID:              1
    ~~~

> 

Reviewer: I think we can add another step here about accessing the sql shell of the single-node server.

```
docker exec -it roach-single ./cockroach sql --url="<sql-address>"
```

where the `sql-address` is the `sql` in the above output (e.g. `postgresql://root@127.0.0.1:26257/defaultdb?sslcert=certs%2Fclient.root.crt&sslkey=certs%2Fclient.root.key&sslmode=verify-full&sslrootcert=certs%2Fca.crt`)

Author: Done. I actually added another "step" to connect to the cluster.

When you use the `cockroach start-single-node` command to start a single-node cluster with Docker, additional features are available to help with testing and development.

{{site.data.alerts.callout_danger}}

> Single-node clusters are not highly-available or fault-tolerant. They are not appropriate for production use.

Reviewer 2: ```suggestion
Single-node clusters are not highly available or fault tolerant. They are not appropriate for production use.
```

{{site.data.alerts.end}}

- You can optionally set the following [Docker environment variables](https://docs.docker.com/engine/reference/commandline/run/#set-environment-variables--e---env---env-file) to create a database and user automatically and to set a password for the user.
  - `COCKROACH_DATABASE`
  - `COCKROACH_USER`
  - `COCKROACH_PASSWORD`

>   To prevent data loss, the environment variables are ignored if the `/cockroach/cockroach-data` directory within the container is not empty. To re-initialize the default database, user, and password from the environment variable values, delete the contents of `/cockroach/cockroach-data` within the running container.

Reviewer 2: I'm concerned about expressing a requirement as a negative (not empty). Could you say if the directory has contents?

Author: Good point, fixed everywhere.

>   To prevent data loss, the environment variables are ignored if the `/cockroach/cockroach-data` directory within the container is not empty. To re-initialize the default database, user, and password from the environment variable values, delete the contents of `/cockroach/cockroach-data` within the running container.

Reviewer 2: The layout of this section is confusing. I think you want to indent the paragraph at line 12 more.
<img width="903" alt="Screen Shot 2022-09-29 at 2 54 57 PM" src="https://user-images.githubusercontent.com/20482960/193149341-afbfbe2a-4092-4a05-8054-1fdc84ea5b3b.png">

Author: Fixed.

> - You can optionally mount a directory of initialization scripts into the `docker-entrypoint-initdb.d` directory within the container. These scripts are run after CockroachDB starts and after the database and user (if specified as environment variables) have been created. The scripts run in the alphanumeric sort order imposed by your locale. The environment variables are ignored if the `/cockroach/cockroach-data` directory within the container is not empty.

Reviewer 2: Same comment about "not empty"

Author: Ack.

This section shows how to start a single-node cluster that uses these features.

### Step 1. Create Docker volumes for the node

Cockroach Labs recommends that you store cluster data in a Docker volume rather than in the storage layer of the running container. Otherwise, if a Docker container is inadvertently deleted, its data is inaccessible.

> To create the [Docker volume](https://docs.docker.com/storage/volumes/) where the cluster will store its data:

Reviewer 2: ```suggestion
To create the [Docker volume](https://docs.docker.com/storage/volumes/) where the cluster will store its data, run the following:
```

{% include_cached copy-clipboard.html %}
~~~ shell
docker volume create roach-single
~~~

> ### Step 2. Start the cluster

Reviewer 2: The formatting of the substeps is broken.

<img width="1088" alt="Screen Shot 2022-09-29 at 2 43 05 PM" src="https://user-images.githubusercontent.com/20482960/193147526-ad0f4130-95a1-4cd8-8063-d296d2c43218.png">

Author: Re-implemented the sub-bullets as HTML and captured the sub-note in a Liquid variable.

1. Start the cluster node.

    The following command starts a single-node cluster that:
 
    - Stores its data in the `roach-single` volume on the Docker host, which is mounted on the `/cockroach/cockroach-data` directory within the container.

>     - If the `/cockroach/cockroach-data` within the container is empty, the specified database, user, and password are created automatically.

Reviewer 2: ```suggestion
    - If the `/cockroach/cockroach-data` within the container is empty, creates the specified database, user, and password automatically.
```

During local development and testing, you can re-initialize the default database, user, and password by deleting the contents of `/cockroach/cockroach-data` within the running container and then restarting the container.

This section shows how to start a single-node cluster that uses these features.

### Step 1. Create a Docker volume for the node

Cockroach Labs recommends that you store cluster data in a Docker volume rather than in the storage layer of the running container. Otherwise, if a Docker container is inadvertently deleted, its data is inaccessible.

To create the [Docker volume](https://docs.docker.com/storage/volumes/) where the cluster will store its data, run the following:

{% include_cached copy-clipboard.html %}
~~~ shell
docker volume create roach-single
~~~

### Step 2. Start the cluster

{% capture  env_file_note %}{{site.data.alerts.callout_success}}Instead of specifying each value directly by using the `-e` or `--env` flag, you can store them in a file on the Docker host. Use one key-value pair per line and set the `--env-file` flag to the file's path.{{site.data.alerts.end}}{% endcapture %}

1. Start the cluster node.

>     The following command starts a single-node cluster that:

Reviewer 2: The command that is "following" is so far away because all the conditions come before it.  Here's an alternative.

To start a cluster that:

bullets

run the following command:

code block

Also it's odd that you are presenting a tip in the middle of the bullets for an alternate way to set the env variables, but then the code snippet disregards that tip. Why not put the tip at the very end of either the bullets or after the code block.  If you pulled the tip out from the bullets you might be able to go back to Markdown instead of HTML.

<img width="1016" alt="Screen Shot 2022-09-30 at 7 11 43 AM" src="https://user-images.githubusercontent.com/20482960/193289396-137aecef-2410-46e8-95e6-9258d0a1ef87.png">

    ~~~

    ~~~ shell
    CockroachDB node starting at 2022-09-28 14:35:27.495508396 +0000 UTC m=+1.370085757 (took 0.5s)
    build:               CCL {{ page.page_version }} @ 2022/09/26 18:49:07 (go1.19.1)
    webui:               https://172.18.0.3:8080
    sql:                 postgresql://root@172.18.0.3:26257/defaultdb?sslcert=certs%2Fclient.root.crt&sslkey=certs%2Fclient.root.key&sslmode=verify-full&sslrootcert=certs%2Fca.crt
    sql (JDBC):          jdbc:postgresql://172.18.0.3:26257/defaultdb?sslcert=certs%2Fclient.root.crt&sslkey=certs%2Fclient.root.key&sslmode=verify-full&sslrootcert=certs%2Fca.crt&user=root
    RPC client flags:    /cockroach/cockroach <client cmd> --host=172.18.0.3:26257 --certs-dir=certs
    logs:                /cockroach/cockroach-data/logs
    temp dir:            /cockroach/cockroach-data/cockroach-temp2611102055
    external I/O path:   /cockroach/cockroach-data/extern
    store[0]:            path=/cockroach/cockroach-data
    storage engine:      pebble
    clusterID:           60f29a4e-1c87-4b0c-805d-eb73460766b1
    status:              initialized new cluster
    nodeID:              1
    ~~~

### Step 3. Connect to the cluster

> 

Reviewer 2: The formatting of some substeps is broken.  The code under substep 2 needs to be indented 1 more space (it's 3 spaces and all the other code snippets are indented 4 spaces) so that that the final step renders as 3.

<img width="1052" alt="Screen Shot 2022-09-30 at 7 05 41 AM" src="https://user-images.githubusercontent.com/20482960/193288092-8c738e11-65e8-4163-8dc3-bda0a3557d4d.png">

When you use the `cockroach start-single-node` command to start a single-node cluster with Docker, additional features are available to help with testing and development.

{{site.data.alerts.callout_danger}}

> Single-node clusters are not highly available or fault tolerant. They are not appropriate for production use.

Reviewer 3: ```suggestion
Single-node clusters are not highly available or fault-tolerant. They are not appropriate for production use.
```

Author: This is funny because Stephanie gave me the exact opposite feedback that resulted in no hyphen here. I can put it back.
